packages/types/src/getters/utils/create-getter.ts

```diff
@@ -0,0 +1,35 @@
+import { Getter } from './getter';
+import { GetterMissingValueError } from './getter-missing-value-error';
+
+export const createGetter = <SourceType, TargetType, Optional extends boolean = false>(
+  getterFunction: (value: SourceType | undefined) => TargetType | undefined,
+  optional?: Optional,
+): Getter<SourceType, TargetType, Optional> => {
+  const getter: Getter<SourceType, TargetType, Optional> = value => {
+    const result = getterFunction(value);
+    if (typeof result === 'undefined' && !optional) {
+      const errorMessage = `Failed to extract from ${JSON.stringify(value)}`;
+      throw new GetterMissingValueError(errorMessage);
+    }
+    return result as Optional extends true ? TargetType | undefined : TargetType;
+  };
+
+  getter.optional = () =>
+    createGetter<SourceType, TargetType, true>(value => getterFunction(value), true);
```

Comment:
it makes sense to me to write Getter as a class or a more detailed object literal, so we can createGetter optional on access, instead of requiring an empty call. this would turn the `.optional()(etc)` syntax into just `.optional(etc)`

```ts
get optional() {
    return createGetter<SourceType, TargetType, true>(value => getterFunction(value), true);
}
```

Reply:
Ah, I see, it's a getter syntax trick